Updated Member model to cascade on destroy

allouis opened the pull request:
no-issue

Up until now we have left orphaned rows in members_stripe_* tables when
a member is deleted, this updates the destroy method so that we cascade
and remove any MemberStripeCustomer and StripeCustomerSubscription
models related to the Member.

This also adds regression tests for the new functionality as well as to
confirm the existing functionality of cascading to the members_labels
join table

allouis committed:
Updated Member model to cascade on destroy

no-issue

Up until now we have left orphaned rows in members_stripe_* tables when
a member is deleted, this updates the destroy method so that we cascade
and remove any MemberStripeCustomer and StripeCustomerSubscription
models related to the Member.

This also adds regression tests for the new functionality as well as to
confirm the existing functionality of cascading to the members_labels
join table

allouis commented on core/server/models/member.js:
```diff
@@ -223,6 +225,51 @@ const Member = ghostBookshelf.Model.extend({
         return options;
+    },
+
+    async destroy(unfilteredOptions) {
```
Kind of annoying that we've essentially duplicated the logic from the BaseModel `destroy` rather than using "super" via the prototype.

The reason for this is that we need the `id` of the model that is going to be deleted, so we can find the customers & subscriptions to delete.

`destroy` takes a `destroyBy` so it could be used to destroy by email, and we would not be passed an `id` in the options.

This means we _have_ to to the whole `forge().fetch()` dance to get the model, and at that point it felt sensible to just add the final `destroy` there too.

ErisDS replied:
What is the reason for doing this explicitly rather than using relation config and allowing bookshelf-relations to handle it?

ErisDS replied:
FWIW this doesn't need to block merge but I would like to understand.

allouis replied:
> What is the reason for doing this explicitly rather than using relation config and allowing bookshelf-relations to handle it?

Both Kevin and I had a call this week to plan our week and both seemed to be under the impression that the relations wouldn't work for this :thinking: Coming to this fresh this morning though that is wrong.

And for the wider question - "how come these relations weren't added originally?" - they were! And then subsequently removed again as they weren't being used and was trying to keep everything as explicit as possible.

The reason they weren't being used is because the webhook handling would/does receive information about a Subscription with no context of a Member, and just `upsert`ing the subscription information. (Rather than have the member context and then attaching the customer/subscription as children).

> this doesn't need to block merge

:+1: I think it is worth me timeboxing an hour this morning though to get this working with relations - that seems much cleaner going foward

ErisDS replied:
Ok cool yes please give it a whirl for an hour as it _would_ unblock several other things :)

allouis replied:
I've pushed some commits with the relations set up :relaxed: 

The initial tests still pass (yay) & I've added some more tests for the relations, it all looks good apart from one thing which I can't get working.

[This test](https://github.com/TryGhost/Ghost/pull/12077/files#diff-844c7ebbc6c104746308a04599f77a6bR16) _should_ pass, but I cannot work out why it doesn't. It means that the relations setup isn't really complete, however the functionality isn't used anywhere as of yet.

I don't want to get stuck in a rabbit hole of debugging it, especially when it isn't used/required - but I think it's not enough of a problem to warrant _not_ using the relations until it's solved, what do you think? I'd be inclined to raise it as a tech debt issue, unless you can see something immediately obvious with fresh eyes?

allouis commented:
Hey @kevinansfield would you mind giving this a quick review, mainly just want to check that this is what you were expecting from the Bulk Delete side, expected usage is shown in the tests :relaxed: 

Would it be preferable to have transactions built it? So that you don't have to manually wrap it in the controller?

allouis committed:
Replaced explicit destroy with relations

allouis committed:
Added "inverse" relations for Stripe models

This adds the relations of Subscription->Customer & Customer->Member

allouis committed:
Added tests for Stripe model relations

allouis committed:
Ensured relations always write in transaction

allouis committed:
Removed explicit transaction from tests

no-issue

This is no longer needed as the models enforce transactions internally
